### PR TITLE
[circle-mpqsolver] Introduce VISQErrorApproximator

### PR DESCRIPTION
**Pull request by stamalakhov**
This commit introduces VISQErrorApproximator.

Draft: https://github.com/Samsung/ONE/pull/10432
Related: https://github.com/Samsung/ONE/issues/10393

ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: compiler/circle-mpqsolver/CMakeLists.txt [hunk 1]
```diff
@@ -1,9 +1,18 @@
+nnas_find_package(Jsoncpp)
+if(NOT Jsoncpp_FOUND)
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/CMakeLists.txt:2]: Lets split this change as this is not directly related with VISQErrorApproximator

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp [hunk 2]
```diff
@@ -73,6 +86,11 @@ int entry(int argc, char **argv)
     .help("Data type of quantized model's outputs (default: uint8)");
 
   arser.add_argument("--output_model").required(true).help("Output quantized model");
+  arser.add_argument("--visq_file")
+    .type(arser::DataType::STR)
+    .default_value("")
+    .required(false)
+    .help("*.visq.json file with quantization errors");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:93]: ditto

File: compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp [hunk 3]
```diff
@@ -1,387 +0,0 @@
-/*
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp:1]: split this removal too

**Comment by jinevening** [comment]
Please address @seanshpark 's comment.

File: compiler/circle-mpqsolver/CMakeLists.txt (hunk 1 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/CMakeLists.txt:2]: ok

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp (hunk 2 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:93]: ok

File: compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp (hunk 3 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp:1]: ok

**Comment by stamalakhov** [comment]
@nnfw-bot test nncc-debug

File: compiler/circle-mpqsolver/CMakeLists.txt (hunk 1 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/CMakeLists.txt:2]: @seanshpark 
This one is needed to build VISQErrorApproximator.

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp (hunk 2 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/CircleMPQSolver.cpp]: done

File: compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp (hunk 3 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp]: done

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  /**
+   * @brief initiliaze by visq_data_path (throws on failure)
+   */
+  void init(const std::string &visq_data_path);
+
+  /**
+   * @brief initiliaze by visq_data (returns success)
+   */
+  bool init(std::istream &visq_data);
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h:44]: Is this method public for testing?

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h:44]: yes. So it can be tested from `VISQErrorApproximator.test.cpp`

**Review by jinevening — DISMISSED** [review]
LGTM

File: compiler/circle-mpqsolver/CMakeLists.txt (hunk 1 above)
Comment (seanshpark) [reply compiler/circle-mpqsolver/CMakeLists.txt:2]: you can split `VISQErrorApproximator` to parts.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/CMakeLists.txt:2]: @seanshpark 
May i  suggest the following sequence of PR's:
1. disabled tests, no `json`
2. `json`'s part and enavling tests
?
Or may be there is another more preferable sequence?

Comment (seanshpark) [reply compiler/circle-mpqsolver/CMakeLists.txt:2]: anyway to meet `introduce VISQErrorApproximator`

Comment (stamalakhov) [reply compiler/circle-mpqsolver/CMakeLists.txt]: @seanshpark 
seems like it's fixed.

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h (hunk 4 above)
Comment (seanshpark) [reply compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h:44]: I don't think adding a method inside `VISQErrorApproximator` just for testing is a good idea.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h]: `init` is moved to private section. tests will be compiled as `friend`

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Review by jinevening — APPROVED** [review]
LGTM

Sorry for late reply.
